### DISCUSSION TITLE
Quantify "DHT Pressure"

### DISCUSSION THREAD

**Discussion by geiseri** [arvidn/libtorrent#6577] [Q&A]
After searching google for some time I have found there have been a lot of conversations with respect to "DHT Pressure" when performing excessive activity on the DHT.  This has been mostly around `get_message` mutable/immutable data, but it has been brought up with `find_node` also.  What I cannot seem to find is a quantification of what the boundaries are.  I assume that DHT operations will take some level of network IO that will grow with traversal.  There would also be CPU cycles on the peers to process those operations.  Is there a way to determine the impact of 100 peers performing mutable data get/put operations on each other?   How does that change if each peer has a mutable entry, and every other peer will query the other single mutable entry?  Likewise, how does that change if only half the peers are looking for another single mutable entry?  Does this grow linearly with the number of peers, or is it exponential/geometric/etc?

TL;DR; How can I figure out the limit of DHT operations before performance deteriorates for peers?

**Comment by arvidn** [comment]
I think the main limiting factor is the rate limit applied to DHT nodes. This defaults to 8 kB/s in libtorrent.
see https://libtorrent.org/reference-Settings.html#dht_upload_rate_limit
The way this rate limit works is that it only applies to *responses*. If the node wants to make requests itself, it's free to do that. But if an incoming request is received, and there's no upload quota left (currently) to send a response, the request is dropped.

i.e. the total number of requests that can be handled (and responded to) is proportional to the total number of nodes on the network (assuming node-IDs and keys are uniformly distributed).
